runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -1226,7 +1227,7 @@ public TimerInternals timerInternals() {
      * fire time of the timer. Used for calculating the output watermark hold. This avoids fetching
      * timer data from the state backend which is expensive if done for each timer.
      */
-    private final PriorityQueue<Long> outputTimestampQueue;
+    private final TreeMultiset<Long> outputTimestamps = TreeMultiset.create();
```

Comment:
We could also consider making this even more efficient by using a `TreeMap<Long, Integer>` where the key is the output timestamp and the value the number of timers which have set it, similar to how it's done in `FlinkStateInternals` for the watermark holds.

Further, we could remove `outputTimestamps` entirely and simply use `stateInternals.addWatermarkHoldUsage(output_timestamp)` and `stateInternals.removeWatermarkHoldUsage(output_timestamp)`.

Reply:
I don't know the internal details of TreeMultiset, but I'd suppose it to be pretty much the same as Map<Long, Integer> in terms of performance. Removing the outputTimestamps entirely seems to be the best option. I'll try that and let you know.

Reply:
Maybe we could replace the TreeMap<Long, Integer> with the TreeMultiset?

Reply:
Why would we do this? TreeMap<Long, Integer> seems like the better data structure because it stores the data compacted.

Reply:
TreeMultiset does the same. And saves some code and performance, the value is stored as array of length 1 and thus adding an already existing item to the set is just get() and in-place increment, while the TreeMap must do get & replace. I think the biggest argument here is that the code is more readable, though. I'll update the PR in a sec, so that you can see.

Reply:
Pushed

Reply:
Wasn't familiar with the internals of TreeMultiSet but if that's the case, it sounds excellent! :)